pkg/controllers/provisioner/v1alpha1/controller.go

```diff
@@ -61,17 +64,48 @@ func (c *Controller) Reconcile(object controllers.Object) error {
 
 	unschedulable := []*v1.Pod{}
 	for _, pod := range pods.Items {
+		tempPod := pod
 		for _, condition := range pod.Status.Conditions {
 			if condition.Type == v1.PodScheduled && condition.Reason == v1.PodReasonUnschedulable {
-				unschedulable = append(unschedulable, &pod)
+				unschedulable = append(unschedulable, &tempPod)
 			}
 		}
 	}
-
-	// 4. Attempt to schedule remaining pods by creating a set of nodes
-	if err := c.Allocator.Allocate(unschedulable); err != nil {
+	pendingPods := c.removeProcessedPods(unschedulable)
```

Comment:
Instead of removing them (mutation), we can orient the code like this:
```
for _, pod := pods {
  if predicate1(pod) &&
    predicate2(pod) &&
    predicate3(pod) {
    provisionable = append(provisionable, pod)
  }
}
allocator.allocate(provisionable)
```
WDYT?

Reply:
I like this approach you have suggested, however, its makes it difficult to remove the pods from cache which are not pending anymore.